src/backend/decayrange.h

```diff
@@ -0,0 +1,290 @@
+     * first place we look for allocating chunks.
+     */
+    static inline // alignas(CACHELINE_SIZE)
+      std::atomic<size_t>
+        epoch{0};
+
+    /**
+     * Flag to ensure one-shot registration with the PAL.
+     */
+    static inline std::atomic_bool registered_timer{false};
+
+    std::atomic_bool registered_local{false};
+
+    /**
+     * All activated DecayRanges.
+     */
+    static inline std::atomic<DecayRange*> all_local{nullptr};
+
+    DecayRange* next{nullptr};
+
+    static void handle_decay_tick()
+    {
+      auto new_epoch = (epoch + 1) % NUM_EPOCHS;
+      // Flush old index for all threads.
+      auto curr = all_local.load(std::memory_order_acquire);
+      while (curr != nullptr)
+      {
+        for (size_t sc = 0; sc < NUM_SLAB_SIZES; sc++)
+        {
+          // Don't use ChunkRecord, store in pagemap.
+          auto old_stack = curr->chunk_stack[sc][new_epoch].pop_all();
+          while (!old_stack.is_empty())
+          {
+            auto next = old_stack.get_next();
+
+            curr->parent->dealloc_range(
+              old_stack.get_capability(), MIN_CHUNK_SIZE << sc);
+
+            old_stack = next;
+          }
+        }
+        curr = curr->next;
+      }
+
+      // Advance current index
+      epoch = new_epoch;
+    }
+
+    class DecayMemoryTimerObject : public PalTimerObject
+    {
+      /***
+       * Method for callback object to perform lazy decommit.
+       */
+      static void process(PalTimerObject*)
+      {
+        handle_decay_tick();
+      }
+
+      // Specify that we notify the ChunkAllocator every 500ms.
+      static constexpr size_t PERIOD = 500;
+
+    public:
+      constexpr DecayMemoryTimerObject() : PalTimerObject(&process, PERIOD) {}
+    };
+
+    static inline DecayMemoryTimerObject timer_object;
+
+  public:
+    class State
+    {
+      DecayRange commit_range{};
+
+    public:
+      constexpr State() = default;
+
+      DecayRange* operator->()
+      {
+        return &commit_range;
+      }
+    };
+
+    static constexpr bool Aligned = ParentRange::Aligned;
+
+    constexpr DecayRange() = default;
+
+    capptr::Chunk<void> alloc_range(size_t size)
+    {
+      // Check local cache
+
+      if constexpr (pal_supports<Time, PAL>)
+      {
+        auto slab_sizeclass = bits::next_pow2_bits(size) - MIN_CHUNK_BITS;
+        // Try local cache of chunks first
+        for (size_t e = 0; e < NUM_EPOCHS; e++)
+        {
+          auto p = chunk_stack[slab_sizeclass][(epoch - e) % NUM_EPOCHS].pop();
+
+          if (p != nullptr)
+            return p;
+        }
+      }
+
+      capptr::Chunk<void> result;
+      for (auto i = NUM_EPOCHS + 2; i > 0; i--)
+      {
+        // Nothing in local cache to allocate from parent.
+        result = parent->alloc_range(size);
+        if (result != nullptr)
+        {
+          return result;
+        }
+
+        // We have run out of memory.
+        handle_decay_tick(); // Try to free some memory.
```

Comment:
Does this need to be interlocked against the timer firing?  I suppose not due to the prepend-only nature of `all_local`, the read-only nature of the spine traversal, and the use of `pop_all` for each found sizeclass... assuming that the `parent` range doesn't need interlocking, which, by default anyway, it doesn't (specifically, the parent will be a `CommitRange` whose parent is a `GlobalRange` by default, and `CommitRange` doesn't actually have state and `GlobalRange` is an interlock).

The presumption of concurrency-safeness of `parent` might merit being written down somewhere?

Reply:
Yeah, I plan to add a static constexpr to all the types that is the concurrency safe like currently happens with Align. I just hadn't threaded it through yet.